### PR TITLE
Improved the reconnect behaviour of CULNetworkHandlerImpl

### PR DESCRIPTION
**Pull request by dereulenspiegel**
This Pull Request addresses the issues described in #4035. I replaced the current implementation of CULNetworkHandlerImpl with a nio based one to simplify handling of connection errors. 
To enable this change slight changes in other components of the CUL transport were necessary. Mainly the responsibility of low level communication is now in the respective handler implementations and not in the AbstractCULHandler any more.

### CODE REVIEW

**Comment by teichsta** [comment]
great thanks @dereulenspiegel!

Could the other CUL users please have a look at this one please?

@cyclingengineer @querdenker2k @johgoe @tarioch

**Comment by tarioch** [comment]
Looks quite good, small suggestion, in CULSerialHandlerImpl, can you remove/make local is und os, they are never accessed directly so therefore they aren't needed.

**Comment by dereulenspiegel** [comment]
InputStream and OutputStream aren't class variables any more.

**Comment by teichsta** [comment]
can any of the others @cyclingengineer @querdenker2k @johgoe confirm this working?

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java [hunk 1]
```diff
@@ -295,21 +262,8 @@ private void requestCreditReport() {
      * @throws CULCommunicationException
      */
     private void writeMessage(String message) throws CULCommunicationException {
-        String deviceName = config.getDeviceAddress();
-        log.debug("Sending raw message to CUL " + deviceName + ":  '" + message + "'");
-        if (bw == null) {
-            log.error("Can't write message, BufferedWriter is NULL");
-        }
-        synchronized (bw) {
```
Comment (johgoe) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java]: New implementation is not synchronized. What happens in parallel write message events?

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java [hunk 2]
```diff
@@ -295,21 +262,8 @@ private void requestCreditReport() {
      * @throws CULCommunicationException
      */
     private void writeMessage(String message) throws CULCommunicationException {
-        String deviceName = config.getDeviceAddress();
-        log.debug("Sending raw message to CUL " + deviceName + ":  '" + message + "'");
-        if (bw == null) {
```
Comment (johgoe) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java]: CULSerialHandlerImpl has no warning for bw == null in log file

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java [hunk 3]
```diff
@@ -210,45 +206,21 @@ protected void notifyError(Exception e) {
...
-                log.warn(
-                        "(LOVF) Limit Overflow: Last message lost. You are using more than 1% transmitting time. Reduce the number of rf messages");
-                return;
-            } else if (data.matches("^\\d+\\s+\\d+")) {
-                processCreditReport(data);
-                return;
-            }
-            notifyDataReceived(data);
-            requestCreditReport();
-        } catch (SocketException e) {
```
Comment (johgoe) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java]: Reopen in case of a SocketException in CULSerialHandlerImpl is missing.

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java [hunk 4]
```diff
@@ -210,45 +206,21 @@ protected void notifyError(Exception e) {
      * @throws CULCommunicationException
      *             if
      */
-    protected void processNextLine() throws CULCommunicationException {
-        String deviceName = config.getDeviceAddress();
-        try {
-            String data = br.readLine();
-            if (data == null) {
```
Comment (johgoe) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java]: EOF Case is missing in CULSerialHandlerImpl

**Comment by johgoe** [comment]
@teichsta, @dereulenspiegel The normal good case with a CUL is working, but I'm not sure about the parallel and error cases. See my comments in the code. 
I guess these features should also move from the base class AbstractCULHandler to the inherited class CULSerialHandlerImpl.

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java (hunk 4 above)
Comment (dereulenspiegel) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java]: @johgoe I don't know what you mean here. processNextLine() was replaced by processNextLine(String), so any I/O error like EOF can't be handled in AbstractCULHandler any more.

**Comment by dereulenspiegel** [comment]
The serial connection part wasn't original part of my changes. It was merely touched because I changed the logic in AbstractCULHandler.
I know added synchronization to BufferedReader and BufferedWriter in the serial implementation and I am now trying to reopen the serial connection on connection errors. However I am not sure if this useful, since a serial connection is vastly different from a network connection on lower levels. My guess is, that every connection error in the serial connection won't be recoverable in software.

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java (hunk 4 above)
Comment (johgoe) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java]: Which part handles EOF instead?

Comment (dereulenspiegel) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/AbstractCULHandler.java]: The specific transport implementations do this as far as possible.

**Comment by 9037568** [comment]
Is there a consensus that this is ready to be merged? ( @dereulenspiegel @tarioch @johgoe @querdenker2k @cyclingengineer )

My review of the changes shows only that the source needs to be run through the eclipse formatter...

**Comment by johgoe** [comment]
From my point of view, its ok to merge this.

**Comment by tarioch** [comment]
Good from my side as well

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/network/CULNetworkHandlerImpl.java [hunk 5]
```diff
@@ -31,58 +37,35 @@
  * (CUN for example).
  *
  * @author Markus Heberling
+ * @author Till Klocke
  * @since 1.5.0
  */
-public class CULNetworkHandlerImpl extends AbstractCULHandler<CULNetworkConfig> {
+public class CULNetworkHandlerImpl extends AbstractCULHandler<CULNetworkConfig>implements Runnable {
```
Comment (watou) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/network/CULNetworkHandlerImpl.java]: a lack of space before `implements` suggests the Eclipse IDE formatting wasn't applied or the settings are incorrect.

**Comment by JSurf** [comment]
I had a chance to test the pull request against my setup (cul connected to ser2net). It seems to work fine and recovers great from ser2net restarts

File: bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/serial/CULSerialHandlerImpl.java [hunk 6]
```diff
@@ -130,6 +142,33 @@ protected void closeHardware() {
...
+            log.error("Failed to reopen serial connection after connection error", e);
+        }
+    }
+
+    @Override
+    protected void write(String command) {
+
+        try {
+            synchronized (bw) {
+                if (bw == null) {
```
Comment (9037568) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/serial/CULSerialHandlerImpl.java]: @watou, would you be ok with merging this PR as is if I submit a separate PR to address the formatting and null synchronization issues?

Comment (watou) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/serial/CULSerialHandlerImpl.java]: @9037568 fine with me. Should this and your follow-on PR target 1.9.1 or is that too risky?

Comment (9037568) [inline bundles/io/org.openhab.io.transport.cul/src/main/java/org/openhab/io/transport/cul/internal/serial/CULSerialHandlerImpl.java]: Probably enough change there to not call it a patch release. Let's call it 1.10.0.
